packages/notifications/push/amplify_push_notifications/android/src/main/kotlin/com/amazonaws/amplify/amplify_push_notifications/PushNotificationBackgroundService.kt

```diff
@@ -0,0 +1,155 @@
+    }
+
+    override fun onCreate() {
+        super.onCreate()
+        startPushNotificationService(this)
+    }
+
+    private fun startPushNotificationService(context: Context) {
+        synchronized(serviceStarted) {
+            if (backgroundFlutterEngine == null) {
+
+                val callbackHandle = context.getSharedPreferences(
+                    PushNotificationConstants.SHARED_PREFERENCES_KEY, Context.MODE_PRIVATE
+                ).getLong(
+                    PushNotificationConstants.CALLBACK_DISPATCHER_HANDLE_KEY, 0
+                )
+                if (callbackHandle == 0L) {
+                    Log.w(TAG, "Warning: no callback registered")
+                    return
+                }
+                val callbackInfo =
+                    FlutterCallbackInformation.lookupCallbackInformation(callbackHandle)
+                if (callbackInfo == null) {
+                    Log.e(TAG, "Error: failed to find callback")
+                    return
+                }
+                Log.i(TAG, "Starting PushNotificationBackgroundService...")
+
+                // Create a background Flutter Engine
+                backgroundFlutterEngine = FlutterEngine(context)
+
+                val args = DartExecutor.DartCallback(
+                    context.assets, FlutterMain.findAppBundlePath(), callbackInfo
+                )
+
+                // DartCallback must only be executed after the engine has been created
+                backgroundFlutterEngine!!.dartExecutor.executeDartCallback(args)
+            }
+        }
+        backgroundChannel = MethodChannel(
+            backgroundFlutterEngine!!.dartExecutor.binaryMessenger,
+            "plugins.flutter.io/amplify_push_notification_plugin_background"
+        )
+        backgroundChannel.setMethodCallHandler(this)
+    }
+
+    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
+        when (call.method) {
+            "PushNotificationBackgroundService.initialized" -> {
+                synchronized(serviceStarted) {
+                    // If events were added to the queue when the service was initializing, emits those
+                    while (!queue.isEmpty()) {
+                        backgroundChannel.invokeMethod("", queue.removeFirst()[0])
+                    }
+
+                    // The background engine has now started and ready to receive events
+                    serviceStarted.set(true)
+                }
+            }
+            else -> result.notImplemented()
+        }
+        result.success(null)
+    }
+
+    override fun onHandleWork(intent: Intent) {
+        Log.i(TAG, "Handling work @ PushNotificationBackgroundService...")
+        if (!intent.isPushNotificationIntent) {
+            return
+        }
+        val remoteMessage = RemoteMessage(intent.extras)
+        val notificationMap = remoteMessage.asPayload().asChannelMap()
+
+        val bgExternalCallbackHandle = baseContext.getSharedPreferences(
```

Comment:
Let's get better names for these - and reduce to a single callback handler if possible

Reply:
There is only one callback function that the customer provides to the background API. 
The handle to that is in `bgExternalCallbackHandle` in the above code, this is sent to the [Callback Dispatcher](https://github.com/aws-amplify/amplify-flutter/blob/base-project-setup-bg/packages/notifications/push/amplify_push_notifications/lib/callback_dispatcher.dart) maintained by the plugin which then calls the customer's callback function.